File: android/src/main/java/com/nitrohealth/WorkoutActivityTypeMapping.kt
package com.nitrohealth

import androidx.health.connect.client.records.ExerciseSessionRecord
import com.margelo.nitro.nitrohealth.NativeWorkoutActivity
import com.margelo.nitro.nitrohealth.NativeWorkoutActivityMapping
import com.margelo.nitro.nitrohealth.NativeWorkoutActivityPortability
import com.margelo.nitro.nitrohealth.NativeWorkoutActivityStatus

/**
 * Maps [ExerciseSessionRecord] exercise type constants to normalized cross-platform activities.
 *
 * The int literals cover exercise type values that connect-client 1.1.0 no longer exposes as
 * `EXERCISE_TYPE_*` constants (they survive only as segment types), but which may still appear
 * in session data written by apps against older schemas. Unknown or future values fall back
 * to an explicit unknown result.
 */
internal fun makeNativeWorkoutActivity(exerciseType: Int): NativeWorkoutActivity {
    val activityType = normalizedWorkoutActivityType(exerciseType)
        ?: return NativeWorkoutActivity(
            status = NativeWorkoutActivityStatus.UNKNOWN,
            type = null,
            portability = null,
            mapping = null
        )

    return NativeWorkoutActivity(
        status = NativeWorkoutActivityStatus.KNOWN,
        type = activityType,
        portability = if (activityType in writableWorkoutActivityTypes) {
            NativeWorkoutActivityPortability.PORTABLE
        } else {
            NativeWorkoutActivityPortability.READONLY
        },
        mapping = if (exerciseType in broadenedHealthConnectExerciseTypes) {
            NativeWorkoutActivityMapping.BROADENED
        } else {
            NativeWorkoutActivityMapping.EXACT
        }
    )
}

private fun normalizedWorkoutActivityType(exerciseType: Int): String? {
    return when (exerciseType) {
        ExerciseSessionRecord.EXERCISE_TYPE_OTHER_WORKOUT -> "other"
        1 -> "coreTraining" // EXERCISE_TYPE_BACK_EXTENSION (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_BADMINTON -> "badminton"
        3 -> "strengthTraining" // EXERCISE_TYPE_BARBELL_SHOULDER_PRESS (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_BASEBALL -> "baseball"
        ExerciseSessionRecord.EXERCISE_TYPE_BASKETBALL -> "basketball"
        6 -> "strengthTraining" // EXERCISE_TYPE_BENCH_PRESS (legacy)
        7 -> "coreTraining" // EXERCISE_TYPE_BENCH_SIT_UP (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_BIKING -> "cycling"
        ExerciseSessionRecord.EXERCISE_TYPE_BIKING_STATIONARY -> "cycling"
        ExerciseSessionRecord.EXERCISE_TYPE_BOOT_CAMP -> "crossTraining"
        ExerciseSessionRecord.EXERCISE_TYPE_BOXING -> "boxing"
        12 -> "calisthenics" // EXERCISE_TYPE_BURPEE (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_CALISTHENICS -> "calisthenics"
        ExerciseSessionRecord.EXERCISE_TYPE_CRICKET -> "cricket"
        15 -> "coreTraining" // EXERCISE_TYPE_CRUNCH (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_DANCING -> "dance"
        17 -> "strengthTraining" // EXERCISE_TYPE_DEADLIFT (legacy)
        18 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_CURL_LEFT_ARM (legacy)
        19 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_CURL_RIGHT_ARM (legacy)
        20 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_FRONT_RAISE (legacy)
        21 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_LATERAL_RAISE (legacy)
        22 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_TRICEPS_EXTENSION_LEFT_ARM (legacy)
        23 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_TRICEPS_EXTENSION_RIGHT_ARM (legacy)
        24 -> "strengthTraining" // EXERCISE_TYPE_DUMBBELL_TRICEPS_EXTENSION_TWO_ARM (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_ELLIPTICAL -> "elliptical"
        ExerciseSessionRecord.EXERCISE_TYPE_EXERCISE_CLASS -> "other"
        ExerciseSessionRecord.EXERCISE_TYPE_FENCING -> "fencing"
        ExerciseSessionRecord.EXERCISE_TYPE_FOOTBALL_AMERICAN -> "americanFootball"
        ExerciseSessionRecord.EXERCISE_TYPE_FOOTBALL_AUSTRALIAN -> "australianFootball"
        30 -> "coreTraining" // EXERCISE_TYPE_FORWARD_TWIST (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_FRISBEE_DISC -> "discSports"
        ExerciseSessionRecord.EXERCISE_TYPE_GOLF -> "golf"
        ExerciseSessionRecord.EXERCISE_TYPE_GUIDED_BREATHING -> "mindAndBody"
        ExerciseSessionRecord.EXERCISE_TYPE_GYMNASTICS -> "gymnastics"
        ExerciseSessionRecord.EXERCISE_TYPE_HANDBALL -> "handball"
        ExerciseSessionRecord.EXERCISE_TYPE_HIGH_INTENSITY_INTERVAL_TRAINING ->
            "highIntensityIntervalTraining"
        ExerciseSessionRecord.EXERCISE_TYPE_HIKING -> "hiking"
        ExerciseSessionRecord.EXERCISE_TYPE_ICE_HOCKEY -> "hockey"
        ExerciseSessionRecord.EXERCISE_TYPE_ICE_SKATING -> "skating"
        40 -> "calisthenics" // EXERCISE_TYPE_JUMPING_JACK (legacy)
        41 -> "jumpRope" // EXERCISE_TYPE_JUMP_ROPE (legacy)
        42 -> "strengthTraining" // EXERCISE_TYPE_LAT_PULL_DOWN (legacy)
        43 -> "calisthenics" // EXERCISE_TYPE_LUNGE (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_MARTIAL_ARTS -> "martialArts"
        ExerciseSessionRecord.EXERCISE_TYPE_PADDLING -> "paddleSports"
        ExerciseSessionRecord.EXERCISE_TYPE_PARAGLIDING -> "paragliding"
        ExerciseSessionRecord.EXERCISE_TYPE_PILATES -> "pilates"
        49 -> "coreTraining" // EXERCISE_TYPE_PLANK (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_RACQUETBALL -> "racquetball"
        ExerciseSessionRecord.EXERCISE_TYPE_ROCK_CLIMBING -> "climbing"
        ExerciseSessionRecord.EXERCISE_TYPE_ROLLER_HOCKEY -> "hockey"
        ExerciseSessionRecord.EXERCISE_TYPE_ROWING -> "rowing"
        ExerciseSessionRecord.EXERCISE_TYPE_ROWING_MACHINE -> "rowing"
        ExerciseSessionRecord.EXERCISE_TYPE_RUGBY -> "rugby"
        ExerciseSessionRecord.EXERCISE_TYPE_RUNNING -> "running"
        ExerciseSessionRecord.EXERCISE_TYPE_RUNNING_TREADMILL -> "running"
        ExerciseSessionRecord.EXERCISE_TYPE_SAILING -> "sailing"
        ExerciseSessionRecord.EXERCISE_TYPE_SCUBA_DIVING -> "underwaterDiving"
        ExerciseSessionRecord.EXERCISE_TYPE_SKATING -> "skating"
        ExerciseSessionRecord.EXERCISE_TYPE_SKIING -> "skiing"
        ExerciseSessionRecord.EXERCISE_TYPE_SNOWBOARDING -> "snowboarding"
        ExerciseSessionRecord.EXERCISE_TYPE_SNOWSHOEING -> "snowSports"
        ExerciseSessionRecord.EXERCISE_TYPE_SOCCER -> "soccer"
        ExerciseSessionRecord.EXERCISE_TYPE_SOFTBALL -> "softball"
        ExerciseSessionRecord.EXERCISE_TYPE_SQUASH -> "squash"
        67 -> "calisthenics" // EXERCISE_TYPE_SQUAT (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_STAIR_CLIMBING -> "stairClimbing"
        ExerciseSessionRecord.EXERCISE_TYPE_STAIR_CLIMBING_MACHINE -> "stairClimbing"
        ExerciseSessionRecord.EXERCISE_TYPE_STRENGTH_TRAINING -> "strengthTraining"
        ExerciseSessionRecord.EXERCISE_TYPE_STRETCHING -> "flexibility"
        ExerciseSessionRecord.EXERCISE_TYPE_SURFING -> "surfing"
        ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_OPEN_WATER -> "swimming"
        ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_POOL -> "swimming"
        ExerciseSessionRecord.EXERCISE_TYPE_TABLE_TENNIS -> "tableTennis"
        ExerciseSessionRecord.EXERCISE_TYPE_TENNIS -> "tennis"
        77 -> "coreTraining" // EXERCISE_TYPE_UPPER_TWIST (legacy)
        ExerciseSessionRecord.EXERCISE_TYPE_VOLLEYBALL -> "volleyball"
        ExerciseSessionRecord.EXERCISE_TYPE_WALKING -> "walking"
        ExerciseSessionRecord.EXERCISE_TYPE_WATER_POLO -> "waterPolo"
        ExerciseSessionRecord.EXERCISE_TYPE_WEIGHTLIFTING -> "strengthTraining"
        ExerciseSessionRecord.EXERCISE_TYPE_WHEELCHAIR -> "wheelchair"
        ExerciseSessionRecord.EXERCISE_TYPE_YOGA -> "yoga"
        else -> null
    }
}

private val broadenedHealthConnectExerciseTypes = setOf(
    1, 3, 6, 7, 12, 15, 17, 18, 19, 20, 21, 22, 23, 24, 30, 40, 41, 42, 43, 49, 67, 77,
    ExerciseSessionRecord.EXERCISE_TYPE_BIKING_STATIONARY,
    ExerciseSessionRecord.EXERCISE_TYPE_BOOT_CAMP,
    ExerciseSessionRecord.EXERCISE_TYPE_EXERCISE_CLASS,
    ExerciseSessionRecord.EXERCISE_TYPE_GUIDED_BREATHING,
    ExerciseSessionRecord.EXERCISE_TYPE_ICE_SKATING,
    ExerciseSessionRecord.EXERCISE_TYPE_PADDLING,
    ExerciseSessionRecord.EXERCISE_TYPE_ROLLER_HOCKEY,
    ExerciseSessionRecord.EXERCISE_TYPE_ROWING_MACHINE,
    ExerciseSessionRecord.EXERCISE_TYPE_RUNNING_TREADMILL,
    ExerciseSessionRecord.EXERCISE_TYPE_STAIR_CLIMBING_MACHINE,
    ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_OPEN_WATER,
    ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_POOL,
    ExerciseSessionRecord.EXERCISE_TYPE_WEIGHTLIFTING
)

private val writableWorkoutActivityTypes = setOf(
    "americanFootball", "australianFootball", "badminton", "baseball", "basketball",
    "boxing", "climbing", "cricket", "crossTraining", "cycling", "dance", "discSports",
    "elliptical", "fencing", "flexibility", "golf", "gymnastics", "handball",
    "highIntensityIntervalTraining", "hiking", "hockey", "martialArts", "mindAndBody",
    "other", "paddleSports", "pilates", "racquetball", "rowing", "rugby", "running",
    "sailing", "skating", "skiing", "snowboarding", "snowSports", "soccer", "softball",
    "squash", "stairClimbing", "strengthTraining", "surfing", "swimming", "tableTennis",
    "tennis", "volleyball", "walking", "waterPolo", "wheelchair", "yoga"
)

internal fun toHealthConnectWorkoutActivityType(activityType: String): Int {
    require(activityType in writableWorkoutActivityTypes) {
        "Unsupported writable workout activity type: $activityType"
    }
    return when (activityType) {
        "americanFootball" -> ExerciseSessionRecord.EXERCISE_TYPE_FOOTBALL_AMERICAN
        "australianFootball" -> ExerciseSessionRecord.EXERCISE_TYPE_FOOTBALL_AUSTRALIAN
        "badminton" -> ExerciseSessionRecord.EXERCISE_TYPE_BADMINTON
        "baseball" -> ExerciseSessionRecord.EXERCISE_TYPE_BASEBALL
        "basketball" -> ExerciseSessionRecord.EXERCISE_TYPE_BASKETBALL
        "boxing" -> ExerciseSessionRecord.EXERCISE_TYPE_BOXING
        "climbing" -> ExerciseSessionRecord.EXERCISE_TYPE_ROCK_CLIMBING
        "cricket" -> ExerciseSessionRecord.EXERCISE_TYPE_CRICKET
        "crossTraining" -> ExerciseSessionRecord.EXERCISE_TYPE_BOOT_CAMP
        "cycling" -> ExerciseSessionRecord.EXERCISE_TYPE_BIKING
        "dance" -> ExerciseSessionRecord.EXERCISE_TYPE_DANCING
        "discSports" -> ExerciseSessionRecord.EXERCISE_TYPE_FRISBEE_DISC
        "elliptical" -> ExerciseSessionRecord.EXERCISE_TYPE_ELLIPTICAL
        "fencing" -> ExerciseSessionRecord.EXERCISE_TYPE_FENCING
        "flexibility" -> ExerciseSessionRecord.EXERCISE_TYPE_STRETCHING
        "golf" -> ExerciseSessionRecord.EXERCISE_TYPE_GOLF
        "gymnastics" -> ExerciseSessionRecord.EXERCISE_TYPE_GYMNASTICS
        "handball" -> ExerciseSessionRecord.EXERCISE_TYPE_HANDBALL
        "highIntensityIntervalTraining" ->
            ExerciseSessionRecord.EXERCISE_TYPE_HIGH_INTENSITY_INTERVAL_TRAINING
        "hiking" -> ExerciseSessionRecord.EXERCISE_TYPE_HIKING
        "hockey" -> ExerciseSessionRecord.EXERCISE_TYPE_ICE_HOCKEY
        "martialArts" -> ExerciseSessionRecord.EXERCISE_TYPE_MARTIAL_ARTS
        "mindAndBody" -> ExerciseSessionRecord.EXERCISE_TYPE_GUIDED_BREATHING
        "other" -> ExerciseSessionRecord.EXERCISE_TYPE_OTHER_WORKOUT
        "paddleSports" -> ExerciseSessionRecord.EXERCISE_TYPE_PADDLING
        "pilates" -> ExerciseSessionRecord.EXERCISE_TYPE_PILATES
        "racquetball" -> ExerciseSessionRecord.EXERCISE_TYPE_RACQUETBALL
        "rowing" -> ExerciseSessionRecord.EXERCISE_TYPE_ROWING
        "rugby" -> ExerciseSessionRecord.EXERCISE_TYPE_RUGBY
        "running" -> ExerciseSessionRecord.EXERCISE_TYPE_RUNNING
        "sailing" -> ExerciseSessionRecord.EXERCISE_TYPE_SAILING
        "skating" -> ExerciseSessionRecord.EXERCISE_TYPE_SKATING
        "skiing" -> ExerciseSessionRecord.EXERCISE_TYPE_SKIING
        "snowboarding" -> ExerciseSessionRecord.EXERCISE_TYPE_SNOWBOARDING
        "snowSports" -> ExerciseSessionRecord.EXERCISE_TYPE_SNOWSHOEING
        "soccer" -> ExerciseSessionRecord.EXERCISE_TYPE_SOCCER
        "softball" -> ExerciseSessionRecord.EXERCISE_TYPE_SOFTBALL
        "squash" -> ExerciseSessionRecord.EXERCISE_TYPE_SQUASH
        "stairClimbing" -> ExerciseSessionRecord.EXERCISE_TYPE_STAIR_CLIMBING
        "strengthTraining" -> ExerciseSessionRecord.EXERCISE_TYPE_STRENGTH_TRAINING
        "surfing" -> ExerciseSessionRecord.EXERCISE_TYPE_SURFING
        "swimming" -> ExerciseSessionRecord.EXERCISE_TYPE_SWIMMING_POOL
        "tableTennis" -> ExerciseSessionRecord.EXERCISE_TYPE_TABLE_TENNIS
        "tennis" -> ExerciseSessionRecord.EXERCISE_TYPE_TENNIS
        "volleyball" -> ExerciseSessionRecord.EXERCISE_TYPE_VOLLEYBALL
        "walking" -> ExerciseSessionRecord.EXERCISE_TYPE_WALKING
        "waterPolo" -> ExerciseSessionRecord.EXERCISE_TYPE_WATER_POLO
        "wheelchair" -> ExerciseSessionRecord.EXERCISE_TYPE_WHEELCHAIR
        "yoga" -> ExerciseSessionRecord.EXERCISE_TYPE_YOGA
        else -> error("Writable workout activity mapping is incomplete for: $activityType")
    }
}

File: ios/WorkoutActivityTypeMapping.swift
import Foundation

enum WorkoutActivityTypeMappingError: LocalizedError {
    case unsupportedWritableType(String)

    var errorDescription: String? {
        switch self {
        case .unsupportedWritableType(let activityType):
            return "Unsupported writable workout activity type: \(activityType)"
        }
    }
}

enum HealthKitWorkoutActivityMapping: Equatable {
    enum Portability: Equatable {
        case portable
        case readOnly
    }

    enum Fidelity: Equatable {
        case exact
        case broadened
    }

    case known(type: String, portability: Portability, mapping: Fidelity)
    case unknown
}

// Switches on raw values so future HealthKit values remain distinguishable from the native
// `.other` case and this semantic mapping stays testable without importing HealthKit.
func makeHealthKitWorkoutActivityMapping(rawValue: UInt) -> HealthKitWorkoutActivityMapping {
    let type: String
    let mapping: HealthKitWorkoutActivityMapping.Fidelity

    switch rawValue {
    case 1: type = "americanFootball" // .americanFootball
    case 2: type = "archery" // .archery
    case 3: type = "australianFootball" // .australianFootball
    case 4: type = "badminton" // .badminton
    case 5: type = "baseball" // .baseball
    case 6: type = "basketball" // .basketball
    case 7: type = "bowling" // .bowling
    case 8: type = "boxing" // .boxing
    case 9: type = "climbing" // .climbing
    case 10: type = "cricket" // .cricket
    case 11: type = "crossTraining" // .crossTraining
    case 12: type = "curling" // .curling
    case 13: type = "cycling" // .cycling
    case 14: type = "dance" // .dance
    case 15: type = "dance" // .danceInspiredTraining (deprecated)
    case 16: type = "elliptical" // .elliptical
    case 17: type = "equestrianSports" // .equestrianSports
    case 18: type = "fencing" // .fencing
    case 19: type = "fishing" // .fishing
    case 20: type = "strengthTraining" // .functionalStrengthTraining
    case 21: type = "golf" // .golf
    case 22: type = "gymnastics" // .gymnastics
    case 23: type = "handball" // .handball
    case 24: type = "hiking" // .hiking
    case 25: type = "hockey" // .hockey
    case 26: type = "hunting" // .hunting
    case 27: type = "lacrosse" // .lacrosse
    case 28: type = "martialArts" // .martialArts
    case 29: type = "mindAndBody" // .mindAndBody
    case 30: type = "mixedCardio" // .mixedMetabolicCardioTraining (deprecated)
    case 31: type = "paddleSports" // .paddleSports
    case 32: type = "other" // .play
    case 33: type = "flexibility" // .preparationAndRecovery
    case 34: type = "racquetball" // .racquetball
    case 35: type = "rowing" // .rowing
    case 36: type = "rugby" // .rugby
    case 37: type = "running" // .running
    case 38: type = "sailing" // .sailing
    case 39: type = "skating" // .skatingSports
    case 40: type = "snowSports" // .snowSports
    case 41: type = "soccer" // .soccer
    case 42: type = "softball" // .softball
    case 43: type = "squash" // .squash
    case 44: type = "stairClimbing" // .stairClimbing
    case 45: type = "surfing" // .surfingSports
    case 46: type = "swimming" // .swimming
    case 47: type = "tableTennis" // .tableTennis
    case 48: type = "tennis" // .tennis
    case 49: type = "trackAndField" // .trackAndField
    case 50: type = "strengthTraining" // .traditionalStrengthTraining
    case 51: type = "volleyball" // .volleyball
    case 52: type = "walking" // .walking
    case 53: type = "waterFitness" // .waterFitness
    case 54: type = "waterPolo" // .waterPolo
    case 55: type = "waterSports" // .waterSports
    case 56: type = "wrestling" // .wrestling
    case 57: type = "yoga" // .yoga
    case 58: type = "barre" // .barre
    case 59: type = "coreTraining" // .coreTraining
    case 60: type = "skiing" // .crossCountrySkiing
    case 61: type = "skiing" // .downhillSkiing
    case 62: type = "flexibility" // .flexibility
    case 63: type = "highIntensityIntervalTraining" // .highIntensityIntervalTraining
    case 64: type = "jumpRope" // .jumpRope
    case 65: type = "kickboxing" // .kickboxing
    case 66: type = "pilates" // .pilates
    case 67: type = "snowboarding" // .snowboarding
    case 68: type = "stairClimbing" // .stairs
    case 69: type = "stepTraining" // .stepTraining
    case 70: type = "wheelchair" // .wheelchairWalkPace
    case 71: type = "wheelchair" // .wheelchairRunPace
    case 72: type = "taiChi" // .taiChi
    case 73: type = "mixedCardio" // .mixedCardio
    case 74: type = "handCycling" // .handCycling
    case 75: type = "discSports" // .discSports
    case 76: type = "fitnessGaming" // .fitnessGaming
    case 77: type = "dance" // .cardioDance
    case 78: type = "dance" // .socialDance
    case 79: type = "pickleball" // .pickleball
    case 80: type = "flexibility" // .cooldown
    case 82: type = "swimBikeRun" // .swimBikeRun
    case 83: type = "other" // .transition
    case 84: type = "underwaterDiving" // .underwaterDiving
    case 3000: type = "other" // .other
    default: return .unknown
    }

    switch rawValue {
    case 15, 20, 30, 32, 33, 50, 60, 61, 68, 70, 71, 77, 78, 80, 83:
        mapping = .broadened
    default:
        mapping = .exact
    }

    let portability: HealthKitWorkoutActivityMapping.Portability =
        (try? healthKitWorkoutActivityRawValue(type)) == nil ? .readOnly : .portable
    return .known(type: type, portability: portability, mapping: mapping)
}

func healthKitWorkoutActivityRawValue(_ activityType: String) throws -> UInt {
    switch activityType {
    case "americanFootball": return 1
    case "australianFootball": return 3
    case "badminton": return 4
    case "baseball": return 5
    case "basketball": return 6
    case "boxing": return 8
    case "climbing": return 9
    case "cricket": return 10
    case "crossTraining": return 11
    case "cycling": return 13
    case "dance": return 77 // .cardioDance
    case "discSports": return 75
    case "elliptical": return 16
    case "fencing": return 18
    case "flexibility": return 62
    case "golf": return 21
    case "gymnastics": return 22
    case "handball": return 23
    case "highIntensityIntervalTraining": return 63
    case "hiking": return 24
    case "hockey": return 25
    case "martialArts": return 28
    case "mindAndBody": return 29
    case "other": return 3000
    case "paddleSports": return 31
    case "pilates": return 66
    case "racquetball": return 34
    case "rowing": return 35
    case "rugby": return 36
    case "running": return 37
    case "sailing": return 38
    case "skating": return 39
    case "skiing": return 60 // .crossCountrySkiing
    case "snowboarding": return 67
    case "snowSports": return 40
    case "soccer": return 41
    case "softball": return 42
    case "squash": return 43
    case "stairClimbing": return 44
    case "strengthTraining": return 20 // .functionalStrengthTraining
    case "surfing": return 45
    case "swimming": return 46
    case "tableTennis": return 47
    case "tennis": return 48
    case "volleyball": return 51
    case "walking": return 52
    case "waterPolo": return 54
    case "wheelchair": return 70 // .wheelchairWalkPace
    case "yoga": return 57
    default: throw WorkoutActivityTypeMappingError.unsupportedWritableType(activityType)
    }
}
